File: ios/Sources/MoveSafe/Engine/ChecklistEngine.swift
import Foundation

enum ChecklistEngineError: Error, LocalizedError {
    case unknownOriginState(String)
    case unknownDestinationState(String)
    case unknownPurpose(String)

    var errorDescription: String? {
        switch self {
        case .unknownOriginState(let id): return "Unknown origin state: \(id)"
        case .unknownDestinationState(let id): return "Unknown destination state: \(id)"
        case .unknownPurpose(let id): return "Unknown purpose: \(id)"
        }
    }
}

enum ChecklistEngine {
    static func generate(
        input: GenerateChecklistInput,
        reference: Date = Date()
    ) throws -> GenerateChecklistOutput {
        guard let originState = States.byId[input.originStateId] else {
            throw ChecklistEngineError.unknownOriginState(input.originStateId)
        }
        guard let destinationState = States.byId[input.destinationStateId] else {
            throw ChecklistEngineError.unknownDestinationState(input.destinationStateId)
        }
        guard let purpose = Purposes.byId[input.purposeId] else {
            throw ChecklistEngineError.unknownPurpose(input.purposeId)
        }

        let selectedCategories = Set(input.selectedCategoryIds)
        let selectedFlags = Set(input.selectedProfileFlags)

        let filtered = ChecklistItems.all.filter { item in
            guard selectedCategories.contains(item.categoryId) else { return false }

            // Purpose match (or item applies to all purposes when list is empty).
            if !item.appliesToPurposes.isEmpty,
               !item.appliesToPurposes.contains(input.purposeId) {
                return false
            }

            // Profile-flag match: if item declares flags, at least one must
            // match, unless the user selected no flags at all (then we keep
            // flag-tagged items in to avoid empty results).
            if !item.appliesToProfileFlags.isEmpty, !selectedFlags.isEmpty {
                let anyMatch = item.appliesToProfileFlags.contains { selectedFlags.contains($0) }
                if !anyMatch { return false }
            }

            // State-specific items (none in MVP seed, but enforced for future data).
            if let originId = item.originStateId, originId != input.originStateId {
                return false
            }
            if let destId = item.destinationStateId, destId != input.destinationStateId {
                return false
            }

            return true
        }

        // Attach sources with computed staleness, resolving any state-bound URLs
        // against the origin / destination states. When a binding resolves to a
        // real state URL (i.e. the state had a non-placeholder value for that
        // field), upgrade the source status to CURRENT — it was verified during
        // the research run that populated those URLs.
        let withSources: [ChecklistItemWithSources] = filtered.map { item in
            let attached = item.sourceIds
                .compactMap { Sources.byId[$0] }
                .map { src -> SourceWithStatus in
                    let resolved = resolveUrl(for: src, origin: originState, destination: destinationState)
                    let baseStatus = SourceStatusEngine.compute(for: src, reference: reference)
                    let didResolveAgainstState = src.stateUrlBinding != nil && resolved != src.url
                    let status: SourceStatus = didResolveAgainstState ? .current : baseStatus
                    return SourceWithStatus(
                        source: src,
                        computedStatus: status,
                        resolvedUrl: resolved
                    )
                }
            return ChecklistItemWithSources(item: item, sources: attached)
        }

        // Group by category. Sort items within each group by display risk order.
        let riskOrderIndex = Dictionary(
            uniqueKeysWithValues: RiskLevel.displayOrder.enumerated().map { ($1, $0) }
        )

        var groupedByCategory: [String: [ChecklistItemWithSources]] = [:]
        for entry in withSources {
            groupedByCategory[entry.item.categoryId, default: []].append(entry)
        }
        for key in groupedByCategory.keys {
            groupedByCategory[key]?.sort { a, b in
                (riskOrderIndex[a.item.riskLevel] ?? 0) < (riskOrderIndex[b.item.riskLevel] ?? 0)
            }
        }

        // Order group keys by category.sortOrder.
        let ordered: [(categoryId: String, items: [ChecklistItemWithSources])] =
            Categories.all
                .compactMap { cat in
                    guard let items = groupedByCategory[cat.id], !items.isEmpty else { return nil }
                    return (cat.id, items)
                }

        // Risk counts across visible items.
        var riskCounts: [RiskLevel: Int] = [
            .verifySeparately: 0, .high: 0, .medium: 0, .low: 0
        ]
        for entry in withSources {
            riskCounts[entry.item.riskLevel, default: 0] += 1
        }

        // Warnings derived from the result set.
        var warnings: [String] = []

        let hasPlaceholder = withSources.contains { entry in
            entry.sources.contains { $0.computedStatus.isPlaceholderOrMissing }
        }
        if hasPlaceholder {
            warnings.append("Some checklist items use placeholder sources. Replace with official sources before relying on them.")
        }

        let hasStale = withSources.contains { entry in
            entry.sources.contains { $0.computedStatus.isStale }
        }
        if hasStale {
            warnings.append("Some sources have not been verified recently. Check the official website before relying on them.")
        }

        if (riskCounts[.verifySeparately] ?? 0) > 0 {
            warnings.append("High-risk categories require separate verification. MoveSafe does not determine whether possession, transportation, or use of regulated items is lawful.")
        }

        warnings.append("This checklist is not a legal conclusion. Verify with official sources and consult a licensed professional where appropriate.")

        if input.originStateId == input.destinationStateId {
            warnings.insert(
                "Origin and destination are the same. MoveSafe is designed for state-to-state moves — some items may not apply.",
                at: 0
            )
        }

        return GenerateChecklistOutput(
            originState: originState,
            destinationState: destinationState,
            purpose: purpose,
            groupedItems: ordered,
            riskCounts: riskCounts,
            warnings: warnings,
            totalItems: withSources.count
        )
    }

    /// Resolves a source's display URL. If the source has a state-binding, pulls
    /// from the relevant state's matching agency URL; falls back to the source's
    /// literal `url` (typically `"#"` for placeholders) when the state has no
    /// value for that field.
    private static func resolveUrl(for source: Source, origin: USState, destination: USState) -> String {
        guard let binding = source.stateUrlBinding else { return source.url }
        let state = (binding.role == .destination) ? destination : origin
        return state.url(for: binding.field) ?? source.url
    }

    /// Unique sources across the grouped output, preserving first-seen order.
    static func uniqueSources(
        from grouped: [(categoryId: String, items: [ChecklistItemWithSources])]
    ) -> [SourceWithStatus] {
        var seen = Set<String>()
        var ordered: [SourceWithStatus] = []
        for (_, items) in grouped {
            for entry in items {
                for s in entry.sources where !seen.contains(s.source.id) {
                    seen.insert(s.source.id)
                    ordered.append(s)
                }
            }
        }
        return ordered
    }
}

File: src/lib/utils/checklistEngine.ts
import { categories, categoriesById } from "../data/categories";
import { checklistItems } from "../data/checklistItems";
import { purposesById } from "../data/purposes";
import { sources, sourcesById } from "../data/sources";
import { statesById } from "../data/states";
import { riskOrder } from "../constants";
import type {
  ChecklistItemWithSources,
  GenerateChecklistInput,
  GenerateChecklistOutput,
  RiskLevel,
  Source,
  SourceStatus,
} from "../types";
import { getSourceStatus, isSourcePlaceholderOrMissing, isSourceStale } from "./sourceStatus";

function attachSources(
  sourceIds: string[],
  reference: Date
): Array<Source & { computedStatus: SourceStatus }> {
  return sourceIds
    .map((id) => sourcesById.get(id))
    .filter((s): s is Source => Boolean(s))
    .map((s) => ({ ...s, computedStatus: getSourceStatus(s, reference) }));
}

export function generateChecklist(
  input: GenerateChecklistInput,
  reference: Date = new Date()
): GenerateChecklistOutput {
  const originState = statesById.get(input.originStateId);
  const destinationState = statesById.get(input.destinationStateId);
  const purpose = purposesById.get(input.purposeId);

  if (!originState) throw new Error(`Unknown origin state: ${input.originStateId}`);
  if (!destinationState)
    throw new Error(`Unknown destination state: ${input.destinationStateId}`);
  if (!purpose) throw new Error(`Unknown purpose: ${input.purposeId}`);

  const selectedCategories = new Set(input.selectedCategoryIds);
  const selectedFlags = new Set(input.selectedProfileFlags);

  // Filter checklist items based on the user's selections.
  const filtered = checklistItems.filter((item) => {
    if (!selectedCategories.has(item.categoryId)) return false;

    // Purpose match (or item applies to all purposes when list is empty).
    if (
      item.appliesToPurposes.length > 0 &&
      !item.appliesToPurposes.includes(input.purposeId)
    ) {
      return false;
    }

    // Profile-flag match: if item declares flags, at least one must match,
    // unless the user selected no flags at all (then we keep flag-tagged
    // items in to avoid empty results when users skip optional flags).
    if (item.appliesToProfileFlags.length > 0 && selectedFlags.size > 0) {
      const anyFlagMatches = item.appliesToProfileFlags.some((f) =>
        selectedFlags.has(f)
      );
      if (!anyFlagMatches) return false;
    }

    // State-specific items (none in MVP seed, but enforced for future data).
    if (item.originStateId && item.originStateId !== input.originStateId) {
      return false;
    }
    if (
      item.destinationStateId &&
      item.destinationStateId !== input.destinationStateId
    ) {
      return false;
    }

    return true;
  });

  // Attach sources with computed staleness.
  const withSources: ChecklistItemWithSources[] = filtered.map((item) => ({
    ...item,
    sources: attachSources(item.sourceIds, reference),
  }));

  // Group by category and sort within group by risk order.
  const grouped: Record<string, ChecklistItemWithSources[]> = {};
  for (const item of withSources) {
    (grouped[item.categoryId] ||= []).push(item);
  }
  for (const id of Object.keys(grouped)) {
    grouped[id].sort(
      (a, b) =>
        riskOrder.indexOf(a.riskLevel) - riskOrder.indexOf(b.riskLevel)
    );
  }

  // Sort category keys by category.sortOrder so the UI is stable.
  const orderedGrouped: Record<string, ChecklistItemWithSources[]> = {};
  for (const cat of categories) {
    if (grouped[cat.id]) orderedGrouped[cat.id] = grouped[cat.id];
  }

  // Risk counts across all visible items.
  const riskCounts: Record<RiskLevel, number> = {
    VERIFY_SEPARATELY: 0,
    HIGH: 0,
    MEDIUM: 0,
    LOW: 0,
  };
  for (const item of withSources) riskCounts[item.riskLevel]++;

  // Warnings derived from the result set.
  const warnings: string[] = [];

  const hasPlaceholder = withSources.some((i) =>
    i.sources.some((s) => isSourcePlaceholderOrMissing(s.computedStatus))
  );
  if (hasPlaceholder) {
    warnings.push(
      "Some checklist items use placeholder sources. Replace with official sources before relying on them."
    );
  }

  const hasStale = withSources.some((i) =>
    i.sources.some((s) => isSourceStale(s.computedStatus))
  );
  if (hasStale) {
    warnings.push(
      "Some sources have not been verified recently. Check the official website before relying on them."
    );
  }

  if (riskCounts.VERIFY_SEPARATELY > 0) {
    warnings.push(
      "High-risk categories require separate verification. MoveSafe does not determine whether possession, transportation, or use of regulated items is lawful."
    );
  }

  if (categoriesById.get("cat-high-risk")?.id && !selectedCategories.has("cat-high-risk")) {
    // Note when high-risk was not selected so the user knows it isn't covered.
  }

  warnings.push(
    "This checklist is not a legal conclusion. Verify with official sources and consult a licensed professional where appropriate."
  );

  if (input.originStateId === input.destinationStateId) {
    warnings.unshift(
      "Origin and destination are the same. MoveSafe is designed for state-to-state moves — some items may not apply."
    );
  }

  return {
    originState,
    destinationState,
    purpose,
    groupedItems: orderedGrouped,
    riskCounts,
    warnings,
    totalItems: withSources.length,
  };
}

// Helper: how many unique sources exist across the result set.
export function getUniqueSources(
  grouped: Record<string, ChecklistItemWithSources[]>
): Array<Source & { computedStatus: SourceStatus }> {
  const map = new Map<string, Source & { computedStatus: SourceStatus }>();
  for (const items of Object.values(grouped)) {
    for (const item of items) {
      for (const s of item.sources) {
        if (!map.has(s.id)) map.set(s.id, s);
      }
    }
  }
  return Array.from(map.values());
}

export { sources };
